### PR TITLE
Use namespaces and accept multiple filters values

### PR DESCRIPTION
**Pull request by fastslack**
Pull Request for Issue # .

### Summary of Changes

This patch use namespaces to load Joomla! libraries and accept multiple values as array for filtering. 

Also add the option to set context value in the xml fields. For example:

```
        <field
                name="article"
                type="sql"
                label="COM_EXAMPLE_FORM_ARTICLE"
                sql_select="c.id, c.title"
                sql_from="#__content AS c"
                sql_order="c.id ASC"
                sql_filter="catid"
                key_field="id"
                value_field="title"
                context="com_example.list_view_example"
        />
```

Into the [documentation](https://docs.joomla.org/Special:MyLanguage/SQL_form_field_type) for custom fields is well explained how to use default filters setting sql_default_{field} and a static value but is not explained how to set dynamic variables using user states for filtering. (see testing)

### Testing Instructions

Using the last field example, we can see that we set **sql_filter** as **catid** . For filtering we must to set into our **view.html.php** file, using the **setUserState** method a value or an array for multiple values. For example:

```
$multiple_values = array(3, 5, 7); 
Factory::getApplication()->setUserState("com_example.list_view_example.filter.catid",  $multiple_values);
```
This is going to filter into the **catid** field using the values declared in the variable **$multiple_values**

### Expected result

We should get the list filtered correctly using the static or dynamic values.

### Documentation Changes Required

Dynamic filtering must to be added to:

https://docs.joomla.org/Special:MyLanguage/SQL_form_field_type

### CODE REVIEW

**Comment by HLeithner** [comment]
Looks like a new feature and have to go into j4, please rebase on j4-dev branch and create a new PR.

**Comment by fastslack** [comment]
Its a fix, not a new feature. Documentation is missing for actual code....

**Comment by alikon** [comment]
@fastslack  is this related to #23568 and / or #22241

**Comment by fastslack** [comment]
@alikon Yes Nicola! Related with both. Fix context is the key to solve this problem.

@HLeithner can you reopen please?

**Comment by HLeithner** [comment]
I cleanup your PR a bit...

File: libraries/joomla/form/fields/sql.php [hunk 1]
```diff
@@ -9,7 +9,12 @@
 
 defined('JPATH_PLATFORM') or die;
 
-JFormHelper::loadFieldClass('list');
+use Joomla\CMS\Factory;
+use Joomla\CMS\Form\FormHelper;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\HTML\HTMLHelper;
```
Comment (Quy) [inline libraries/joomla/form/fields/sql.php:14]: Please sort in alpha order. Move line 15 above line 14.

File: libraries/joomla/form/fields/sql.php [hunk 2]
```diff
@@ -226,21 +234,24 @@ protected function processQuery($conditions, $filters, $defaults)
...
-					$escape = $db->quote($db->escape($html_filters[$value]), false);
-
-					$query->where("{$value} = {$escape}");
+					if (is_array($html_filters[$value]))
+					{
+						$query->where($db->quoteName($value) . ' IN (' . implode(',', $db->quote($html_filters[$value])) . ')');
+					}
+					else
+					{
+						$query->where($db->quoteName($value) . " = " .$db->quote($html_filters[$value]));
```
Comment (Quy) [inline libraries/joomla/form/fields/sql.php]:
```suggestion
						$query->where($db->quoteName($value) . ' = ' .$db->quote($html_filters[$value]));
```

File: libraries/joomla/form/fields/sql.php [hunk 3]
```diff
@@ -226,21 +234,24 @@ protected function processQuery($conditions, $filters, $defaults)
...
+					}
 				}
 				elseif (!empty($defaults[$value]))
 				{
-					$escape = $db->quote($db->escape($defaults[$value]), false);
-
-					$query->where("{$value} = {$escape}");
+					$query->where($db->quoteName($value) . " = " .$db->quote($defaults[$value]));
```
Comment (Quy) [inline libraries/joomla/form/fields/sql.php]:
```suggestion
					$query->where($db->quoteName($value) . ' = ' .$db->quote($defaults[$value]));
```

File: libraries/joomla/form/fields/sql.php [hunk 4]
```diff
@@ -226,21 +234,24 @@ protected function processQuery($conditions, $filters, $defaults)
 		// Process the filters
 		if (is_array($filters))
 		{
-			$html_filters = JFactory::getApplication()->getUserStateFromRequest($this->context . '.filter', 'filter', array(), 'array');
+			$html_filters = Factory::getApplication()->getUserState($this->context . '.filter', 'filter');
```
Comment (HLeithner) [inline libraries/joomla/form/fields/sql.php:237]: Can the user still change the filter?

Comment (fastslack) [reply libraries/joomla/form/fields/sql.php:237]: Yes, I tested it and seems that work better using **getUserState**. User can change the filter in their views using:
```
$app->setUserState("com_example.list_view_example.filter.catid", $filterCategories);
```

If we use **getUserStateFromRequest**, when the use select one item from the fields, it cleanup all filters and do not get the current filtering. It return empty values

**Comment by Quy** [comment]
Should namespaces be in a separate PR for J4 and focus on multiple filters values only in this PR for J3?

**Comment by fastslack** [comment]
Should I? I can make a separate PR but I think tjhat Joomla! 3.9 use namespaces too

**Comment by Quy** [comment]
Preferably one PR per issue to keep it simple for testing.

**Comment by hakanara** [comment]
Dear all,

I really appreciate the efforts you put into fixing the dynamic filtering feature of SQL field. But, there is still no documentation update for this feature in Joomla docs. I read the explanations here to apply and use this feature but view.html.php file does not exist for Joomla menu settings or for a plugin. So how or where can I use this feature? Does it really work?

So please, either update the documentation or please check if this feature works in a plugin or in Joomla menu settings.

I can open a new issue if required. 

Kind regards

**Comment by ghost** [comment]
@hakanara yes, please open a new Issue and link also to this Pull Request, thanks.

**Comment by Quy** [comment]
Reopened #22241

**Pull request closed without merging**